File: 39_combination_sum/combinationSum.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candi, int target) {
        sort(candi.begin(), candi.end());
        vector<vector<int> > res;
        dfs(res, target, candi, vector<int>());
        return res;
    }
private:
    void dfs(vector<vector<int> >&  res, int remain, vector<int>& candi, vector<int> v) {
        if(remain == 0) {
            res.push_back(v);
            return;
        }

        for(int i = 0; i < candi.size() && candi[i] <= remain; i++) {
            if(!v.empty() && candi[i] < v.back())
                continue;
            else {
                v.push_back(candi[i]);
                dfs(res, remain - candi[i], candi, v);
                v.pop_back();
            }
        }
    }
};

int main(int argc, char const *argv[])
{
	Solution sol;
	vector<vector<int> > result;
	int target = 7;
	int arr[] = {2, 3, 4, 7};
	vector<int> nums(arr, arr + (sizeof(arr) / sizeof(*arr)));
	result = sol.combinationSum(nums, target);
	//cout << result.size() << endl;

	for(int j = 0; j < result.size(); ++j)
	{
		for(int i = 0; i < result[j].size(); ++i)
		{
			cout << result[j][i] << " ";
		}
		cout << endl;
	}

	//cout << result[0].size() << endl;
	return 0;
}

File: 39_combination_sum/combinationSum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        result = []
        # If we sort the list, onece we meet an item greater
        # than current remain, we know the remains of the list
        # would larger than the remain. So, we break the for loop.
        candidates = sorted(candidates)
        
        def dfs(remain, stack):
            if not remain:
                result.append(stack)
                return 

            for item in candidates:
                if item > remain: break
                # To avoid duplicate result, we have to avoid push
                # the item that is smaller than the last item of the stack
                # With this if condition, we can get non-duplicated stacks 
                # with increasing order. 
                if stack and item < stack[-1]: continue
                else:
                    dfs(remain - item, stack + [item])
        
        dfs(target, [])
        return result


candidates = [2, 3, 6, 7]
target = 100
sol = Solution()
res = sol.combinationSum(candidates, target)
print res
